components/datetime/src/raw/neo.rs

```diff
@@ -0,0 +1,72 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::calendar::CldrCalendar;
+use crate::pattern::runtime;
+use icu_plurals::PluralCategory;
+use icu_provider::prelude::*;
+use zerovec::ZeroMap;
+use crate::provider::neo::*;
+use crate::format::neo::*;
+
+pub(crate) enum DatePatternSelectionData {
+    SingleDate(DataPayload<ErasedDatePatternV1Marker>),
+    OptionalEra {
+        with_era: DataPayload<ErasedDatePatternV1Marker>,
+        without_era: DataPayload<ErasedDatePatternV1Marker>,
+    },
+}
+
+pub(crate) enum Foo1<'data> {
+    SingleDate(runtime::Pattern<'data>),
+    WeekPlurals(ZeroMap<'data, PluralCategory, runtime::PatternULE>),
+}
+
+pub(crate) struct Foo2<'data> {
+    map: ZeroMap<'data, PluralCategory, runtime::PatternULE>,
+}
+
+pub(crate) enum TimePatternData {
+    SingleTime(DataPayload<TimePatternV1Marker>),
+}
+
+pub(crate) enum DateTimePatternSelectionData {
+    Date(DatePatternSelectionData),
+    Time(TimePatternData),
+    Glue(DataPayload<DateTimePatternV1Marker>),
+}
+
+pub(crate) struct TypedNeoDateTimeFormatter<C: CldrCalendar> {
+    names: TypedDateTimeNames<C>,
+}
+
+pub(crate) struct RawNeoDateTimeFormatter {
+    names: RawDateTimeNames,
+    selection: DateTimePatternSelectionData,
+}
+
+#[cfg(test)]
+mod tests {
+    use icu_calendar::week::WeekCalculator;
+    use icu_decimal::FixedDecimalFormatter;
+
+    use super::*;
+
+    #[test]
+    fn test_sizes() {
+        assert_eq!(48, core::mem::size_of::<DataPayload<ErasedDatePatternV1Marker>>());
+        assert_eq!(80, core::mem::size_of::<DataLocale>());
+        assert_eq!(216, core::mem::size_of::<FixedDecimalFormatter>());
+        assert_eq!(2, core::mem::size_of::<WeekCalculator>());
+        assert_eq!(72, core::mem::size_of::<DataPayload<ErasedYearNamesV1Marker>>());
+        assert_eq!(56, core::mem::size_of::<DataPayload<ErasedMonthNamesV1Marker>>());
+        assert_eq!(40, core::mem::size_of::<DataPayload<WeekdayNamesV1Marker>>());
+        assert_eq!(40, core::mem::size_of::<DataPayload<DayPeriodNamesV1Marker>>());
```

Comment:
Thought: We could save space in DateTimeNames by making this `LiteMap<Field, DataNamesPayload, ShortVec>` where `DataNamesPayload` is an enum between the payload types. It compels an allocation however for a pattern containing more than 1 display name. @Manishearth

Reply:
Most patterns need more than one display name, right?

Reply:
Well patterns containing only AM/PM ("9:46 pm" or "12/26/2023 at 9:46 pm") or only the month name ("Dec 26, 2023") are fairly common, but yes for anything more than that we would allocate which isn't great. But, saving stack space is nice. I don't want to do this now, but something to at least consider in #4337